sdk/go/pulumi-language-go/main.go

```diff
@@ -339,10 +337,38 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
 		return nil, errors.Wrap(err, "failed to prepare environment")
 	}
 
-	cmd, err := findProgram(host.binary)
+	// the user can explicitly opt in to using a binary executable by specifying
+	// runtime.options.binary in the Pulumi.yaml
+	program := host.binary
+	if program == "" {
+		// user did not specify a binary and we will compile and run the binary on-demand
+		logging.V(5).Infof("No prebuilt executable specified, attempting invocation via compilation")
+
+		pulumiHome, err := workspace.GetPulumiHomeDir()
+		if err != nil {
+			return nil, err
+		}
+
+		file, err := ioutil.TempFile(pulumiHome, ".pulumi.out")
```

Comment:
Deprecated: As of Go 1.17, this function simply calls os.CreateTemp.

I think based on our discussion it should be "" indicating system temp folder, not pulumiHome; hogging ~/.pulumi/ with 100s of temp files is not great; we could use a subfolder like `path.Join(pulumiHome, "go-compiler-cache")` but then we'd need to include commands to manage/prune that into Pulumi; since that's currently out of scope perhaps just system TEMP is best.

Reply:
I'd recommend just using TEMP for now. As I said on the issue maybe there's something clever we could do here to benefit from compiler caching but I wouldn't worry about that for a first pass (I don't think "go run" is benefiting from it?)